File: 2021/08/07/remove_linked_list_elements.py
"""
Problem: https://leetcode.com/problems/reverse-bits/
Author: Andrew Jarombek
Date: 8/6/2021
"""

# Ran 1:16:19 Half-Marathon.  Super tired and stomach isn't happy with me now, but happy with the race.

from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def remove_elements(head: Optional[ListNode], val: int) -> Optional[ListNode]:
    if head is None:
        return None

    new_head = head
    current = head
    previous = None

    while current is not None:
        if current.val == val:
            if previous is None:
                new_head = current.next
                current = new_head
            else:
                previous.next = current.next
                current = current.next
        else:
            previous = current
            current = current.next

    return new_head

File: 2022/03/28/c/removeLinkedListElements.c
// Program: https://leetcode.com/problems/remove-linked-list-elements/
// Author: Andrew Jarombek
// Date: 3/28/2022

#include <stdlib.h>
#include <assert.h>

struct ListNode {
    int val;
    struct ListNode *next;
};

struct ListNode* removeElements(struct ListNode* head, int val){
    while (head != NULL && head->val == val)
        head = head->next;

    if (head == NULL)
        return NULL;

    struct ListNode* node = head;

    while (node->next != NULL) {
        if (node->next->val == val) {
            node->next = node->next->next;
        } else {
            node = node->next;
        }
    }

    return head;
}

int main() {
    struct ListNode node7 = {6, NULL};
    struct ListNode node6 = {5, &node7};
    struct ListNode node5 = {4, &node6};
    struct ListNode node4 = {3, &node5};
    struct ListNode node3 = {6, &node4};
    struct ListNode node2 = {2, &node3};
    struct ListNode node1 = {1, &node2};

    struct ListNode* result = removeElements(&node1, 6);

    assert(result->val == 1);
    assert(result->next->val == 2);
    assert(result->next->next->val == 3);
    assert(result->next->next->next->val == 4);
    assert(result->next->next->next->next->val == 5);
    assert(result->next->next->next->next->next == NULL);
}
